crypto/s2n_ecc_evp.c

```diff
@@ -0,0 +1,390 @@
+
+static EVP_PKEY *
+s2n_ecc_evp_generate_own_key(const struct s2n_ecc_named_curve *named_curve) {
+  EVP_PKEY *evp_pkey = NULL;
+  EVP_PKEY_CTX *pctx = NULL;
+  EVP_PKEY_CTX *kctx = NULL;
+  EVP_PKEY *params = NULL;
+
+#if S2N_OPENSSL_VERSION_AT_LEAST(1, 1, 0) && !defined(LIBRESSL_VERSION_NUMBER)
+  if (named_curve->libcrypto_nid == NID_X25519) {
+    pctx = EVP_PKEY_CTX_new_id(named_curve->libcrypto_nid, NULL);
+    if (pctx == NULL) {
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    if (EVP_PKEY_keygen_init(pctx) != 1) {
+      EVP_PKEY_CTX_free(pctx);
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    if (EVP_PKEY_keygen(pctx, &evp_pkey) != 1) {
+      EVP_PKEY_CTX_free(pctx);
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+
+    EVP_PKEY_CTX_free(pctx);
+    return evp_pkey;
+  }
+#endif
+  if (named_curve->libcrypto_nid == NID_X9_62_prime256v1 ||
+      named_curve->libcrypto_nid == NID_secp384r1) {
+    pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+    if (pctx == NULL) {
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    if (EVP_PKEY_paramgen_init(pctx) != 1) {
+      EVP_PKEY_CTX_free(pctx);
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(
+            pctx, named_curve->libcrypto_nid) != 1) {
+      EVP_PKEY_CTX_free(pctx);
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    if (!EVP_PKEY_paramgen(pctx, &params)) {
+      EVP_PKEY_CTX_free(pctx);
+      EVP_PKEY_free(params);
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    kctx = EVP_PKEY_CTX_new(params, NULL);
+    if (kctx == NULL) {
+      EVP_PKEY_CTX_free(pctx);
+      EVP_PKEY_free(params);
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    if (EVP_PKEY_keygen_init(kctx) != 1) {
+      EVP_PKEY_CTX_free(pctx);
+      EVP_PKEY_free(params);
+      EVP_PKEY_CTX_free(kctx);
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    if (EVP_PKEY_keygen(kctx, &evp_pkey) != 1) {
+      EVP_PKEY_CTX_free(pctx);
+      EVP_PKEY_free(params);
+      EVP_PKEY_CTX_free(kctx);
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    EVP_PKEY_CTX_free(pctx);
+    EVP_PKEY_CTX_free(kctx);
+    EVP_PKEY_free(params);
+    return evp_pkey;
+  } else {
+    S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+  }
+}
+
+static int s2n_ecc_evp_compute_shared_secret(EVP_PKEY *own_key,
+                                             EVP_PKEY *peer_public,
+                                             struct s2n_blob *shared_secret) {
+  EVP_PKEY_CTX *ctx = NULL;
+  size_t shared_secret_size;
+
+  ctx = EVP_PKEY_CTX_new(own_key, NULL);
+  if (ctx == NULL) {
+    S2N_ERROR(S2N_ERR_ECDHE_SHARED_SECRET);
+  }
+
+  if (EVP_PKEY_derive_init(ctx) != 1) {
+    EVP_PKEY_CTX_free(ctx);
+    S2N_ERROR(S2N_ERR_ECDHE_SHARED_SECRET);
+  }
+  if (EVP_PKEY_derive_set_peer(ctx, peer_public) != 1) {
+    EVP_PKEY_CTX_free(ctx);
+    S2N_ERROR(S2N_ERR_ECDHE_SHARED_SECRET);
+  }
+  if (EVP_PKEY_derive(ctx, NULL, &shared_secret_size) != 1) {
```

Comment:
is this line necessary? if the purpose is to check shared_secret_size, perhaps add a comment

Reply:
will add comments